README.md
# k6 Prometheus Exporter

This is a attempt to provide data from the [k6 load testing tool](https://k6.io/) to Prometheus in order to display and visualize it in Grafana

It is written in Go and acts as proxy for the k6 REST API, exposing the results over HTTP in the standard Prometheus exposition format, at the standard `/metrics` endpoint

Goals:

- See if it was possible using the k6 REST API
- Report on k6 load tests without needing to push huge volumes of data into a CSV or InfluxDB

![](https://img.shields.io/github/license/benc-uk/k6-prometheus-exporter)
![](https://img.shields.io/github/last-commit/benc-uk/k6-prometheus-exporter)
![](https://img.shields.io/github/release/benc-uk/k6-prometheus-exporter)
![](https://img.shields.io/github/checks-status/benc-uk/k6-prometheus-exporter/main)
![](https://img.shields.io/github/workflow/status/benc-uk/k6-prometheus-exporter/CI%20Build?label=ci-build)
![](https://img.shields.io/github/workflow/status/benc-uk/k6-prometheus-exporter/Release%20Assets?label=release)

# Getting Started

## Deploying to Kubernetes

See [example-job.yaml](deploy/example-job.yaml) for an example of how to deploy k6-prometheus-exporter as a sidecar alongside a k6 load test, running as a Kubernetes job.

You can also run it as a standalone pod external from k6, this requires that the k6 REST API (port 6565 by default) is exposed as a service. See [deploy/standalone](./deploy/standalone) for an example of doing this

## Local Development

- If you wish to build or run locally, use `make build` or `make run`
- If you wish to build the Docker image, use `make image` or `make push`. When calling make set **IMAGE_REG**, **IMAGE_REPO** and **IMAGE_TAG** as you wish

## Running as container

```bash
docker run ghcr.io/benc-uk/k6-prometheus-exporter:latest
```

# Screenshots

Example of some of the data shown in a Grafana dashboard

![](./docs/screen2.png)

# Configuration

| Setting / Variable | Purpose                                     | Default                    |
| ------------------ | ------------------------------------------- | -------------------------- |
| METRICS_PORT       | Port the server will listen on.             | 2112                       |
| K6_API_ENDPOINT    | Override the endpoint used to connect to k6 | `http://localhost:6565/v1` |

# Repository Structure

A brief description of the top-level directories of this project is as follows:

```c
/api        - Details of the API specification & docs
/build      - Build configuration e.g. Dockerfiles
/charts     - Helm charts
/deploy     - Deployment and infrastructure as code, inc Kubernetes
/scripts    - Bash and other supporting scripts
/src        - Source code
/test       - Testing, mock data and API + load tests
```

# Known Issues

List any known bugs or gotchas.

# Change Log

See [complete change log](./CHANGELOG.md)

# License

This project uses the MIT software license. See [full license file](./LICENSE)

# Acknowledgements

Optional. Put acknowledgements and credits here, if any


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

Nothing

## [1.0.0] - 2078-06-20

### Added

- Blah blah
- Foo bar

### Changed

- Start using "changelog" over "change log" since it's the common usage.

### Removed

- Nothing


cmd/main.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"sort"
	"strings"
	"time"
)

// Metric from k6 API
type Metric struct {
	ID         string
	Attributes struct {
		Type     string
		Contains string
		Sample   map[string]float64
	}
}

// Group from k6 API
type Group struct {
	ID         string
	Attributes struct {
		Path   string
		Name   string
		Checks []Check
	}
}

// Check from k6 API
type Check struct {
	Name   string
	Passes int
	Fails  int
}

var k6EndpointMetrics = ""
var k6EndpointGroups = ""

func metricsHandler(resp http.ResponseWriter, req *http.Request) {
	// Call the k6 metrics API
	apiRes, err := http.Get(k6EndpointMetrics)
	if err != nil || apiRes.StatusCode != 200 {
		var errorMsg string
		if err != nil {
			errorMsg = err.Error()
		} else {
			errorMsg = apiRes.Status
		}
		log.Printf("Error with downstream k6 API %s\n", errorMsg)

		// We return 200 and empty page on error, why?
		// It stops Prometheus from thinking we're dead and stopping scraping
		_, _ = resp.Write([]byte(""))
		return
	}

	// Hold API result for metrics
	metricData := struct {
		Data []Metric
	}{}

	// JSON unmarshall
	err = json.NewDecoder(apiRes.Body).Decode(&metricData)
	if err != nil {
		resp.WriteHeader(500)
		fmt.Fprintln(resp, err.Error())
		return
	}

	sort.Slice(metricData.Data, func(i, j int) bool {
		return metricData.Data[i].ID < metricData.Data[j].ID
	})

	// Call the k6 groups API
	apiRes, err = http.Get(k6EndpointGroups)
	if err != nil || apiRes.StatusCode != 200 {
		var errorMsg string
		if err != nil {
			errorMsg = err.Error()
		} else {
			errorMsg = apiRes.Status
		}
		log.Printf("Error with downstream k6 API %s\n", errorMsg)

		// We return 200 and empty page on error, why?
		// It stops Prometheus from thinking we're dead and stopping scraping
		_, _ = resp.Write([]byte(""))
		return
	}

	// Hold API result for groups
	groupData := struct {
		Data []Group
	}{}

	// JSON unmarshall
	err = json.NewDecoder(apiRes.Body).Decode(&groupData)
	if err != nil {
		resp.WriteHeader(500)
		fmt.Fprintln(resp, err.Error())
		return
	}

	// Now build Prometheus exposition format
	expoResult := ""

	// Push metrics and sample values into results
	for _, metric := range metricData.Data {
		metricName := metric.ID

		for sampleName, sampleValue := range metric.Attributes.Sample {
			sampleName = strings.ReplaceAll(sampleName, "(", "")
			sampleName = strings.ReplaceAll(sampleName, ")", "")
			expoResult += fmt.Sprintf("k6_%s_%s %f\n", metricName, sampleName, sampleValue)
		}

		expoResult += "\n"
	}

	// Push group checks passes and fails
	for _, group := range groupData.Data {
		groupName := group.Attributes.Name
		if strings.TrimSpace(groupName) == "" {
			groupName = "default"
		}

		for _, check := range group.Attributes.Checks {
			expoResult += fmt.Sprintf("k6_group_check_fail{group_name=\"%s\", check_name=\"%s\"} %d\n", groupName, check.Name, check.Fails)
			expoResult += fmt.Sprintf("k6_group_check_passes{group_name=\"%s\", check_name=\"%s\"} %d\n", groupName, check.Name, check.Passes)
		}

		expoResult += "\n"
	}

	// Return as HTTP response
	_, _ = resp.Write([]byte(expoResult))
}

func main() {
	serverPort := "2112"
	if portEnv := os.Getenv("METRICS_PORT"); portEnv != "" {
		serverPort = portEnv
	}

	k6ApiEndpoint := "http://localhost:6565/v1"
	if endpoint := os.Getenv("K6_API_ENDPOINT"); endpoint != "" {
		k6ApiEndpoint = endpoint
	}
	k6EndpointMetrics = k6ApiEndpoint + "/metrics"
	k6EndpointGroups = k6ApiEndpoint + "/groups"

	http.HandleFunc("/metrics", metricsHandler)

	log.Printf("### k6 metrics proxy listening on %v\n", serverPort)
	log.Printf("### API endpoint for k6: %s\n", k6ApiEndpoint)
	srv := &http.Server{
		Addr:         fmt.Sprintf(":%s", serverPort),
		WriteTimeout: 10 * time.Second,
		ReadTimeout:  10 * time.Second,
		IdleTimeout:  10 * time.Second,
	}

	err := srv.ListenAndServe()
	if err != nil {
		panic(err.Error())
	}
}
